File: WebGL/NodeJS/public/src/Cam.js
var camInstanceCount = 0;

class Cam {
	
	constructor()
	{
	}
	
	init(screenW, screenH, zNear, zFar, zNearPlaneW) {
		
		camInstanceCount++;
		
		if(camInstanceCount > 1) {
			throw "More than one object creation is not allowed for Cam Class";
		}
		
		this.projMat = new GLMat();
		this.modelMat = new GLMat();
		
		this.sw = screenW;
		this.sh = screenW;
				
		this._zNear  = zNear;
		this._zFar = zFar;
		this._zNearPlaneHalfW = zNearPlaneW/2.0;

		this._pivot = new CVector3(0, 0, 0);
		this._trans = new CVector3(0, 0, -70.0);
		this._angle = new CVector3(30, 0, 0);

		this.setPerspectiveView();		
	}
	
	setPerspectiveView()
	{
		this._left = -this._zNearPlaneHalfW;
		this._right = this._zNearPlaneHalfW;
		this._bottom = -this._zNearPlaneHalfW*this.sh/this.sw;
		this._top = this._zNearPlaneHalfW*this.sh/this.sw;

		this.projMat.glLoadIdentity();
		this.projMat.glFrustum(this._left, this._right, this._bottom, this._top, this._zNear, this._zFar);
	}
	
	setModelViewMatrix()
	{
		this.modelMat.glLoadIdentity();
		this.modelMat.glTranslatef(this._trans.x, this._trans.y, this._trans.z);
		this.modelMat.glRotatef(this._angle.x,1,0,0);
		this.modelMat.glRotatef(this._angle.y,0,1,0);
		this.modelMat.glTranslatef(-this._pivot.x, -this._pivot.y, -this._pivot.z);
	}
		
	updateCamera()
	{
		if(input.IsKeyPressed(input.shift) && input.IsMiddleMousePressed())
		{
			var dx = input.mouseX.curr - input.mouseX.prev;
			var dy = input.mouseY.curr - input.mouseY.prev;

			var z = this._trans.z;
			if(z < 0)
				z = -z;

			z /= 3000.0;

			this._trans.x += dx*z;
			this._trans.y += -dy*z;

			return true;
		}
		else if(input.IsKeyPressed(input.ctrl) && input.IsMiddleMousePressed())
		{
			this._trans.z += (input.mouseY.prev - input.mouseY.curr) * 2.0;
			return true;
		}
		else if(input.IsMiddleMousePressed())
		{
			var dx = (input.mouseX.curr - input.mouseX.prev);
			var dy = (input.mouseY.curr - input.mouseY.prev);

			this._angle.y += dx * 180.0 / (this.sw*0.5);
			this._angle.x += dy * 180.0 / (this.sw*0.5);

			return true;
		}

		if(input.IsScrollDown())
		{
			this._trans.z -= 45.0;
			return true;
		}
		else if(input.IsScrollUp())
		{
			this._trans.z += 45.0;
			return true;
		}

		return false;
	}
	
}

var cam = new Cam();

File: PhysicsRel/BulletPhysicsIntegratoin/source/Cam.cpp
#include "Cam.h"
#include "Input.h"
#include <math.h>

Cam* Cam::_ref = NULL;

Cam::Cam()
{
}

Cam* Cam::GetInstance()
{
	if(_ref == NULL)
	{
		_ref = new Cam();
	}

	return _ref;
}

void Cam::DeleteInstance()
{
	if(_ref)
	{
		delete _ref;
		_ref = NULL;
	}
}

void Cam::Init(int screenW, int screenH, float zNear, float zFar, float zNearPlaneW)
{
	SW = (float)screenW;
	SH = (float)screenH;

	_zNear  = zNear;
	_zFar = zFar;
	_zNearPlaneHalfW = zNearPlaneW/2.0f;

	_pivot = CVector3(0, 0, 0);
	_trans = CVector3(0, 0, -200.0f);
	_angle = CVector3(30, 0, 0);

	_viewType = 5;

	glViewport(0, 0, (int)SW, (int)SH);
	SetPerspectiveView();
}


void Cam::SetProjection()
{
	if(_isOrtho)
	{
		SetOrthoView();
	}
	else
	{
		SetPerspectiveView();
	}
}


void Cam::SetPerspectiveView()
{
	_isOrtho = false;

	_left = -_zNearPlaneHalfW;
	_right = _zNearPlaneHalfW;
	_bottom = -_zNearPlaneHalfW*SH/SW;
	_top = _zNearPlaneHalfW*SH/SW;

	glMatrixMode(GL_PROJECTION);
	glLoadIdentity();
	glFrustum(_left, _right, _bottom, _top, _zNear, _zFar);
}


void Cam::SetOrthoView()
{
	_isOrtho = true;

	_left = -(-_trans.z *_zNearPlaneHalfW) / _zNear;
	_right = (-_trans.z *_zNearPlaneHalfW) / _zNear;
	_bottom = -((-_trans.z *_zNearPlaneHalfW) / _zNear) * (SH/SW);
	_top = ((-_trans.z *_zNearPlaneHalfW) / _zNear) * (SH/SW);

	glMatrixMode(GL_PROJECTION);
	glLoadIdentity();
	glOrtho(_left, _right, _bottom, _top, _zNear, _zFar);	
}


bool Cam::IsOrthoView()
{
	return _isOrtho;
}


void Cam::SetModelViewMatrix()
{
	glMatrixMode(GL_MODELVIEW);
	glLoadIdentity();
	
	glTranslatef(_trans.x, _trans.y, _trans.z);

	//glTranslatef(_pivot.x, _pivot.y, _pivot.z);

	glRotatef(_angle.x,1,0,0);
	glRotatef(_angle.y,0,1,0);

	glTranslatef(-_pivot.x, -_pivot.y, -_pivot.z);
}


bool Cam::UpdateCamera()
{
	if( Input::IsKeyPressed(VK_SHIFT) && Input::IsMiddleMousePressed())
	{
		float dx = (float)(Input::MX - Input::PrevMX);
		float dy = (float)(Input::MY - Input::PrevMY);

		float z = _trans.z;
		if(z < 0)
			z = -z;

		z /= 3000;

		_trans.x += dx*z;
		_trans.y += -dy*z;

		return true;
	}
	else if(Input::IsKeyPressed(VK_CONTROL) && Input::IsMiddleMousePressed())
	{
		_trans.z += (float)(Input::PrevMY - Input::MY) * 2.0f;

		return true;
	}
	else if(Input::IsMiddleMousePressed())
	{
		float dx = (float)(Input::MX - Input::PrevMX);
		float dy = (float)(Input::MY - Input::PrevMY);

		_angle.y += dx * 180.0f / (SW*0.5f);
		_angle.x += dy * 180.0f / (SH*0.5f);

		return true;
	}

	if(Input::IsScrollDown())
	{
		_trans.z -= 45.0f;
		return true;
	}
	else if(Input::IsScrollUp())
	{
		_trans.z += 45.0f;
		return true;
	}

	return false;
}

void Cam::SetPivot(CVector3 pivotPoint)
{
	_pivot = pivotPoint;
}

void Cam::SetTrans(CVector3 trans)
{
	_trans = trans;
}

void Cam::SetRot(CVector3 rot)
{
	_angle = rot;
}

void Cam::SetFrontView()
{
	_angle.Set(0,0,0);
	_viewType = 0;
}

void Cam::SetBackView()
{
	_angle.Set(0,180,0);
	_viewType = 1;
}

void Cam::SetLeftView()
{
	_angle.Set(0,90,0);
	_viewType = 2;
}

void Cam::SetRightView()
{
	_angle.Set(0,-90,0);
	_viewType = 3;
}

void Cam::SetTopView()
{
	_angle.Set(90,0,0);
	_viewType = 4;
}

void Cam::SetBottomView()
{
	_angle.Set(-90,0,0);
	_viewType = 5;
}

void Cam::ChangeView()
{
	_viewType++;

	if(_viewType >= 6)
		_viewType = 0;

	if(_viewType == 0) SetFrontView();
	else if(_viewType == 1) SetBackView();
	else if(_viewType == 2) SetLeftView();
	else if(_viewType == 3) SetRightView();
	else if(_viewType == 4) SetTopView();
	else if(_viewType == 5) SetBottomView();
}
